src/Profiling.cpp

```diff
@@ -20,23 +24,154 @@ class InjectProfiling : public IRMutator {
+        bool on_stack;
+        Expr size;
+    };
+
+    Scope<AllocSize> func_alloc_sizes;
+
+    // Strip down the tupple name, e.g. f.0 into f
+    string normalize_name(const string& name) {
+        vector<string> v = split_string(name, ".");
+        internal_assert(v.size() > 0);
+        return v[0];
+    }
+
+    int get_func_id(const string& name) {
+        string norm_name = normalize_name(name);
+        int idx = -1;
+        map<string, int>::iterator iter = indices.find(norm_name);
         if (iter == indices.end()) {
             idx = (int)indices.size();
-            indices[op->name] = idx;
+            indices[norm_name] = idx;
         } else {
             idx = iter->second;
         }
+        return idx;
+    }
+
+    Expr compute_allocation_size(const vector<Expr> &extents,
+                                 const Expr& condition,
+                                 const Type& type,
+                                 const std::string &name,
+                                 bool& on_stack) {
+        on_stack = true;
+        int32_t constant_size = Allocate::constant_allocation_size(extents, name);
+        if (constant_size > 0) {
+            int64_t stack_bytes = constant_size * type.bytes();
+            if (stack_bytes > ((int64_t(1) << 31) - 1)) { // Out of memory
+                return 0;
+            } else if (can_allocation_fit_on_stack(stack_bytes)) { // Allocation on stack
+                return Expr((int32_t)stack_bytes);
+            }
+        }
+
+        // Check that the allocation is not scalar (if it were scalar
+        // it would have constant size).
+        internal_assert(extents.size() > 0);
+
+        on_stack = false;
+        Expr size = extents[0];
+        for (size_t i = 1; i < extents.size(); i++) {
+            size *= extents[i];
+        }
+        size = simplify(Select::make(condition, size * type.bytes(), 0));
+        return size;
+    }
+
+    void visit(const Allocate *op) {
+        int idx = get_func_id(op->name);
+
+        vector<Expr> new_extents;
+        bool all_extents_unmodified = true;
+        for (size_t i = 0; i < op->extents.size(); i++) {
+            new_extents.push_back(mutate(op->extents[i]));
+            all_extents_unmodified &= new_extents[i].same_as(op->extents[i]);
+        }
+        Expr condition = mutate(op->condition);
+
+        bool on_stack;
+        Expr size = compute_allocation_size(new_extents, condition, op->type, op->name, on_stack);
+        func_alloc_sizes.push(op->name, {on_stack, size});
+
+        if (!is_zero(size) && on_stack) {
+            const int64_t *int_size = as_const_int(size);
+            internal_assert(int_size != NULL); // Stack size is always a const int
+            func_stack_current[idx] += *int_size;
+            func_stack_peak[idx] = std::max(func_stack_peak[idx], func_stack_current[idx]);
+            debug(1) << "  Allocation on stack: " << op->name << "(" << size << ") in pipeline " << pipeline_name
```

Comment:
debug(1) made sense while developing this, but is probably too verbose to check in for this purpose. For debugging the insides of lowering stages, use debug(3).